### PR TITLE
data loader for reading comprehension task

### PR DESCRIPTION
**Pull request by songfeng**
added doc2dial data loader and dummy data for reading comprehension task.

### CODE REVIEW

**Review by lhoestq — COMMENTED** [review]
Nice ! 

It looks like the dummy data zip file is quite big though (almost 8MB).
Could you try to reduce its size please ?
Ideally it should be under 20KB

**Comment by songfeng** [comment]
Thank you for all the feedback! I have updated the dummy data with a zip under 30KB, which needs to include at least one data instance from both document domain and dialog domain. Please let me know if it is still too big. Thanks!

File: datasets/doc2dial/doc2dial.py [hunk 1]
```diff
@@ -163,6 +212,133 @@ def _split_generators(self, dl_manager):
...
+                                turn_to_predict["reference"], doc["spans"], doc["doc_text"]
+                            )
+                        if not turn_to_predict["references"]:
+                            qa["is_impossible"] = True
+                            qa["answers"] = []
+                        else:
+                            qa["is_impossible"] = False
+                            qa["answers"] = turn_to_predict["references"]
+                            if len(qa["answers"]) < 1:
+                                assert len((qa["answers"])) >= 1
```
Comment (lhoestq) [inline datasets/doc2dial/doc2dial.py:331]: What is this used for ?

File: datasets/doc2dial/doc2dial.py [hunk 2]
```diff
@@ -163,6 +212,133 @@ def _split_generators(self, dl_manager):
...
+
+    def _load_doc_data_rc(self, filepath):
+        doc_filepath = os.path.join(os.path.dirname(filepath), "..", "doc2dial_doc.json")
+        with open(doc_filepath, encoding="utf-8") as f:
+            data = json.load(f)["doc_data"]
+        return data
+
+    def _get_start_end_candidates_rc(self, tss):
+        start_candidates, end_candidates = [], []
+        for ts_id, ts in tss.items():
```
Comment (lhoestq) [inline datasets/doc2dial/doc2dial.py:248]: What do `ts` and `tss` mean here ?

**Review by lhoestq — COMMENTED** [review]
Thanks for reducing the size of the dummy data :) 

Can you add some comments and docstrings to explain your processing please ?

Also can you add an example from the dataset in the `Data Instances` section of the readme of this dataset ?
You can get one by using `load_dataset` and taking the first sample `dataset["train"][0]` for example.
And can you briefly describe the data fields in the `Data Fields` section ?

**Comment by songfeng** [comment]
Thank you again for the feedback! I am not too sure what the preferable style for data instance in readme, but still added my edits. Thanks!

File: datasets/doc2dial/doc2dial.py [hunk 3]
```diff
@@ -411,9 +341,62 @@ def _generate_examples(self, filepath):
...
+                                        turn_to_predict["reference"], doc["spans"], doc["doc_text"]
+                                    )
+                                if not turn_to_predict["references"]:
+                                    qa["is_impossible"] = True
+                                else:
+                                    qa["is_impossible"] = False
+                                    qa["answers"] = turn_to_predict["references"]
+                                    """ Ensure the answers are not empty if the question is answerable."""
+                                    if len(qa["answers"]) < 1:
+                                        assert len((qa["answers"])) >= 1
```
Comment (lhoestq) [inline datasets/doc2dial/doc2dial.py:401]:
```suggestion
                                    assert (
                                        len((qa["answers"])) >= 1
                                    ), "Ensure the answers are not empty if the question is answerable"
```

**Review by lhoestq — APPROVED** [review]
Thank you !

**Pull request merged**